packages/nimble-components/src/table/specs/data-hierarchy-hld.md

```diff
@@ -0,0 +1,168 @@
+# Data Hierarchy
+
+## Problem Statement
+
+There are scenarios where clients want to provide collapsible hierarchy to the rows being presented in the table, where both parent and children rows provide values in typically at least a subset of the same columns.
+
+For example, consider a table of geographic information that included location, population, flag, and possibly other location-related information. The location column could be either a continent, country, state, or any sub-regional construct. It would be reasonable to want to present this information in a way that allowed a user to collapse/expand any outer regional grouping, which itself would contain all rows whose locations existed inside it (e.g. 'United States' would be parent of all rows whose location was a U.S. state).
+
+### Lazy Loading
+
+In addition to just supporting data hierarchy with a fully provided set of data, we also need to support the scenario where a row is presented as a parent, but none of its children are currently represented in the data. Upon expanding the row, the client is then expected to load all of the data for that parent and provide it to the table (via the standard `setData` method).
+
+## Links To Relevant Work Items and Reference Material
+
+-   [#890: Design for hierarchical data support in tables](https://github.com/ni/nimble/issues/890)
+-   [#861: Table hierarchical data support](https://github.com/ni/nimble/issues/861)
+-   [Prototype using flat list that includes a parentId](https://github.com/ni/nimble/tree/data-hierarchy-flat-list-prototype) with [Storybook](https://60e89457a987cf003efc0a5b-qzwoshcidz.chromatic.com/?path=/story/incubating-table--table&args=data:LargeDataSet)
+-   [Prototype with user data being hierarchical](https://github.com/ni/nimble/tree/data-hierarchy-prototype) with [Storybook](https://60e89457a987cf003efc0a5b-yncupvnoes.chromatic.com/?path=/story/incubating-table--table&args=data:LargeDataSet)
+
+## Implementation / Design
+
+Users will provide hierarchical data as a flat list (just as they do for non-hierarchical data), but will have to include a field in their data that is representative of the "parent Id" for that record of data.
+
+### `Table` API
+
+```ts
+public Table() {
+    ...
+    // This attribute determines which field in the record specifies the id that is the
+    // parent row. If it is not set, the row has no parent.
+    @attr({ attribute: 'parent-id-field-name' })
+    public parentIdFieldName?: string;
+
+    // This field would be an option for how we allow a user to specify the field
+    // name where they will provide a value indicating whether or not that row
+    // is a parent row. See the below discussion on lazy-loading. If it is not set,
+    // the row is presumed child-less (and thus will show no expand/collapse button).
+    @attr({ attribute: 'has-children-field-name' })
```

Comment:
Should we follow a model like the row selection api? It's a similar kind of concept, marking certain rows a certain way out-of-band with the rest of the actual dataset.

Reply:
Let me see if I understand the proposed workflow. Would it be something like this:
1. Client calls `setData` on table where some rows will eventually be parents (table has no attributes for flagging a row that it will be a parent, and thus rows won't show an expand/collapse button at this point).
2. Client calls something like `setLazyLoadedParents(rowIds)`. This will then mark the provided rows as parents, which will result in them showing expand/collapse buttons.
3. Same workflow described already once a user expands a parent row.

Pros:
- Clients no longer have to provide an extra field in their data to represent this state, which is pretty nice since that state only has transient value.

Cons:
- Possible UI oddity in that rows will have the expand/collapse button "pop in" to existence.

If I have it right, I think I might agree with you in that I would prefer the UI hiccup to making clients provide an extra field in their data that most rows don't need.